File: Java/LinkedList/03_Singly_Circular_LinkedList_Java/SinglyCL_004.java
/*
===============================================================================
Program Name    : Creation of Singly Circular Linked List using Head Pointer
Program Number  : 004
Description     : This program demonstrates the creation of a Singly Circular
                  Linked List using a head pointer in Java. Three nodes are
                  created dynamically and connected in a circular manner where
                  the last node points back to the head node. The program
                  displays the node data and the references stored in each
                  next pointer.

Date            : July 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Class and Object
2. Default Constructor
3. Self-Referential Class
4. Dynamic Memory Allocation using new
5. Head Pointer
6. Circular Linked List
7. Reference Variables

Key Learning Points:
- The head pointer stores the address of the first node.
- Every node contains data and a reference to the next node.
- The last node points back to the head node.
- A circular linked list has no NULL at the end.
- Printing references verifies the circular connections.

===============================================================================
*/
// Program 04 : Singly Circular Linked List using Head Pointer

class node
{
    public int data;
    public node next;

    public node()
    {
        data = 0;
        next = null;
    }
}

class SinglyCL4
{
    public static void main(String A[])
    {
        node head = null;
        node second = null;
        node third = null;

        // Allocate memory
        head = new node();
        second = new node();
        third = new node();

        // Store data
        head.data = 11;
        second.data = 21;
        third.data = 51;

        // Connect nodes
        head.next = second;
        second.next = third;
        third.next = head;

        // Display data
        System.out.println(head.data);
        System.out.println(second.data);
        System.out.println(third.data);

        // Display next references
        System.out.println(head.next);
        System.out.println(second.next);
        System.out.println(third.next);
    }
}
/*
===============================================================================
Expected Output:
===============================================================================

11
21
51
node@xxxxxxxx
node@xxxxxxxx
node@xxxxxxxx

(Note: The hexadecimal values will differ for every execution because they
represent the object's hash code.)

===============================================================================
Memory Representation:
===============================================================================

                      head
                       │
                       ▼
              +----+------+     +----+------+     +----+------+
              | 11 |   •------->| 21 |   •------->| 51 |   •----+
              +----+------+     +----+------+     +----+------+   |
                   ^                                              |
                   |______________________________________________|

                11 → 21 → 51
                 ↑          │
                 └──────────┘

===============================================================================
Explanation:
===============================================================================

The program defines a self-referential class named node containing two data
members: data and next.

Three node objects are created dynamically using the default constructor.

The variable head acts as the head pointer and always stores the reference of
the first node in the linked list.

The nodes are connected as follows:

head → second → third → head

The last node points back to the head node, forming a circular linked list.
Unlike a linear linked list, there is no NULL reference at the end.

The program prints the data stored in each node and also displays the object
references stored in their next pointers to verify the circular links.

===============================================================================
Program Flow:
===============================================================================

Start Program
      │
      ▼
Create Head and Other Node References
      │
      ▼
Create Three Node Objects
      │
      ▼
Assign Data to Each Node
      │
      ▼
Connect Nodes Circularly
      │
      ▼
Display Node Data
      │
      ▼
Display Next References
      │
      ▼
Program Terminates

===============================================================================
Operations Performed:
===============================================================================

1. Defined a self-referential node class.
2. Declared head, second and third reference variables.
3. Created three node objects dynamically.
4. Assigned data values to each node.
5. Connected the nodes to form a circular linked list.
6. Printed the data stored in each node.
7. Displayed the references stored in the next pointers.

===============================================================================
Time Complexity:
===============================================================================

Node Creation    : O(1)
Node Linking     : O(1)
Display Data     : O(1)

Space Complexity : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 004 demonstrates the creation of a Singly Circular Linked List using
a head pointer. The head pointer always refers to the first node, while the
last node points back to the head, forming a closed circular structure. This
serves as the foundation for implementing traversal, insertion, and deletion
operations in a Singly Circular Linked List.

===============================================================================
*/

File: C++/LinkedList/03_Singly_Circular_Linear_LinkedList_CPP/SinglyCL_004.cpp
/*
===============================================================================
Program Name    : Singly Circular Linked List - Creation of Circular Linked List
Program Number  : 004
Description     : This program demonstrates the creation of a Singly Circular
                  Linked List using three dynamically allocated nodes. The
                  last node is connected back to the first node, forming a
                  circular linked list. It also displays the data stored in
                  each node along with the addresses stored in their next
                  pointers.

Date            : July 2026
Language        : C++
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Dynamic Memory Allocation
2. Self-Referential Structure
3. Pointer Declaration
4. Circular Node Linking
5. Pointer Traversal
6. Address Representation

Key Learning Points:
- Three nodes are created dynamically using the new operator.
- The last node points back to the first node.
- A Singly Circular Linked List has no NULL at the end.
- Each next pointer stores the address of the next node.
- Pointer addresses help visualize the circular connection.

===============================================================================
*/

#include<iostream>
using namespace std;

#pragma pack(1)

struct node
{
    int data;
    struct node *next;
};

typedef struct node NODE;
typedef struct node *PNODE;

int main()
{
    PNODE first = new NODE;
    PNODE second = new NODE;
    PNODE third = new NODE;

    first->data = 11;
    first->next = second;

    second->data = 21;
    second->next = third;

    third->data = 51;
    third->next = first;

    cout<<"Data stored in the nodes:\n"<<endl;

    cout<<first->data<<endl;
    cout<<second->data<<endl;
    cout<<third->data<<endl;

    cout<<"\nAddresses stored in next pointers:\n"<<endl;

    cout<<first->next<<endl;
    cout<<second->next<<endl;
    cout<<third->next<<endl;

    delete first;
    delete second;
    delete third;

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

Data stored in the nodes:

11
21
51

Addresses stored in next pointers:

0x61ff10
0x61ff20
0x61ff00

(Note: Memory addresses will vary for every execution.)

===============================================================================
Memory Representation:
===============================================================================

                      first
                        │
                        ▼
          +------------+------------+
          | data = 11  | next ------|--------------------+
          +------------+------------+                    |
                                                        |
                                                        ▼
                              +------------+------------+
                              | data = 21  | next ------|--------------+
                              +------------+------------+              |
                                                                       |
                                                                       ▼
                                                   +------------+------------+
                                                   | data = 51  | next ------|
                                                   +------------+------------+
                                                            │
                                                            │
                                                            └──────────────► first

Circular Link:

first → second → third
  ▲                    │
  └────────────────────┘

===============================================================================
Pointer Explanation:
===============================================================================

first
└── Stores the address of the first node.

second
└── Stores the address of the second node.

third
└── Stores the address of the third node.

first->next
└── Points to second.

second->next
└── Points to third.

third->next
└── Points back to first, creating the circular link.

===============================================================================
Program Flow:
===============================================================================

Create Three Nodes
        │
        ▼
Store Data
        │
        ▼
Connect First → Second
        │
        ▼
Connect Second → Third
        │
        ▼
Connect Third → First
        │
        ▼
Display Node Data
        │
        ▼
Display Pointer Addresses
        │
        ▼
Delete Nodes
        │
        ▼
Program Ends

===============================================================================
Time Complexity:
===============================================================================

Node Creation       : O(1)
Display Data        : O(1)
Display Addresses   : O(1)

Space Complexity    : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 004 demonstrates how three dynamically allocated nodes are connected
to form a Singly Circular Linked List. Unlike a Singly Linear Linked List,
the last node does not point to NULL; instead, it points back to the first
node, creating a circular structure. Displaying the addresses stored in the
next pointers helps visualize how the circular connection is established.

===============================================================================
*/
